spec.bs

```diff
@@ -972,14 +751,351 @@ To <dfn>generate reporting URLs</dfn> given an {{AuctionAdConfig}} |config| and
+of the following global objects:
+
+ * {{InterestGroupBiddingScriptRunnerGlobalScope}}
+ * {{InterestGroupScoringScriptRunnerGlobalScope}}
+ * {{InterestGroupReportingScriptRunnerGlobalScope}}
+
+<div algorithm>
+  To <dfn>evaluate a bidding script</dfn> given a [=string=] |script| and a [=list=] of arguments
+  |arguments|:
+
+    1. Return the result of [=evaluating a script=] with
+       {{InterestGroupBiddingScriptRunnerGlobalScope}}, |script|, "`generateBid`", and |arguments|.
+</div>
+
+<div algorithm>
+  To <dfn>evaluate a scoring script</dfn> given a [=string=] |script| and a [=list=] of arguments
+  |arguments|:
+
+    1. Return the result of [=evaluating a script=] with
+       {{InterestGroupScoringScriptRunnerGlobalScope}}, |script|, "`scoreAd`", and |arguments|.
+</div>
+
+<div algorithm>
+  To <dfn>evaluate a reporting script</dfn> given a [=string=] |script|, a [=string=]
+  |functionName|, and a [=list=] of arguments |arguments|:
+
+    1. Return the result of [=evaluating a script=] with
+       {{InterestGroupReportingScriptRunnerGlobalScope}}, |script|, |functionName|, and |arguments|.
+</div>
+
+<br>
+
+<div algorithm>
+  To <dfn>create a new script runner agent</dfn>, run these steps:
+
+    1. Let |signifier| be a new unique internal value.
+
+    1. Let |candidateExecution| be a new [=ECMAScript/candidate execution=].
+
+    1. Return a new [=ECMAScript/agent=] whose \[[CanBlock]] is false, \[[Signifier]] is
+       |signifier|, \[[CandidateExecution]] is |candidateExecution|, and \[[IsLockFree1]],
+       \[[IsLockFree2]], and \[[LittleEndian]] are set at the implementation's discretion.
+
+  Note: This algorithm is almost identical to [[HTML]]'s [=create an agent=] algorithm, with the
+  exception that we do not give the returned agent a new [=event loop=], since it does not process
+  [=tasks=] within [=task sources=] in the usual way.
+</div>
+
+<div algorithm>
+  To <dfn>obtain a script runner agent</dfn>, run these steps:
+
+    1. Let |agentCluster| be a  new [=ECMAScript/agent cluster=].
+
+    1. Let |agent| be the result of [=creating a new script runner agent=].
+
+    1. Add |agent| to |agentCluster|.
+
+    1. Return |agent|.
+</div>
+
+<div algorithm>
+  To <dfn>evaluate a script</dfn> with a global type |globalType|, [=string=] |script|, [=string=]
+  |functionName|, and a [=list=] |arguments|, run these steps. They return a [=ECMAScript/Completion
+  Record=], which is either an [=abrupt completion=] (in the case of a parse failure or execution
+  error), or a normal completion populated with the [=ECMAScript/ECMAScript language value=] result
+  of invoking |functionName|.
+
+    1. [=Assert=] that these steps are running [=in parallel=].
+
+    1. Let |agent| be the result of [=obtaining a script runner agent=] given null, true, and
+       false. Run the rest of these steps in |agent|.
+
+       Issue: This exclusively creates a new [=ECMAScript/agent cluster=] for the given |script| to

```

Comment:
How does the implementation work here? Do we use the same global or just the same process? If we're given a script URL, when do we decide *not* to fetch that URL and instead pick an existing global (or process maybe, I'm not sure how the grouping works)?

Reply:
I believe we use the site-isolation logic to pick the process, which means that on desktop systems it's generally one process per interest group owner or seller.
Same global scope depends on execution-mode:
**compatibility mode** - We use new JavaScript execution contexts and global scopes for each invocation.
**group-by-origin mode** - We use the same JavaScript execution context and global scope (only executed once) for all interest groups joined on one site with the same owner.
**frozen-context mode** - We use the same JavaScript execution context and global scope (only executed once, then deep-freezed) for all interest groups with the same owner.

Reply:
I'm wondering if we can separate these ten steps into a separate "prepare script runner" section and then our logic depending on execution-mode can call it in different places?

Reply:
Hmm I think that would work IIUC, but I'm tempted to do this outside of this PR (since I'll be out for almost a week), and land this in the meantime. Does that sound OK?